### PR TITLE
Changes to allow finetuning at scale

### PR DESCRIPTION
**Pull request by peterdays**
- Updated to python3.11
- Updated to pyproject.toml
- Created conda environment.yml for reproducibility
- Data generation:
    - changed strategy of splitting (train, val and test vs train and val)
    - added option to pass a custom function to process the reads
    - added verbose control to finetune_generator
    - added progress bars to data generation
- Finetuning:
    - added progress bars
    - separated gradient accumulation from evaluation

### CODE REVIEW

File: src/methylbert/data/finetune_data_generate.py [hunk 1]
```diff
@@ -138,36 +145,41 @@ def _get_methylseq(dmr, bam_file_path: str, k: int, methyl_caller: str):
...
-        split_ratio: float = 1.0,
+        split_ratios: List[float] = [0.8, 0.1, 0.1],
         n_dmrs: int = -1,
         n_cores: int = 1,
         seed: int = 950410,
         ignore_sex_chromo: bool = True,
-        methyl_caller: str = "bismark"
+        methyl_caller: str = "bismark",
+        verbose: int = 2,
+        read_extract_sequences_func: Optional[callable] = None
```
Comment (hanyangii) [inline src/methylbert/data/finetune_data_generate.py:179]: Is this an option for your own function?

File: src/methylbert/trainer.py [hunk 2]
```diff
@@ -578,18 +586,24 @@ def _iteration(self, steps, data_loader):
...
-
+                    steps_progress_bar.set_postfix(eval_loss=eval_loss)
                     # Reset prediction result
                     del train_prediction_res
                     train_prediction_res =  {"dmr_label":[], "pred_ctype_label":[], "ctype_label":[]}
-             
+
+                self.step += 1
+                duration=0
+                global_step_loss = 0
```
Comment (hanyangii) [inline src/methylbert/trainer.py:596]: I am a bit confused here. Wouldn't the global_step_loss need to be reset when the gradient accumulation step is achieved? I think this should be under "`if (local_step+1) % self._config.gradient_accumulation_steps == 0`". Maybe I should add one more line to divide `global_step_loss` with `gradient_accumulation_step` before recording the performance in `f_perform` file.

**Review by hanyangii — CHANGES_REQUESTED** [review]
I made some comments on your changes! Please let me know if you have different opinions.

File: src/methylbert/data/finetune_data_generate.py [hunk 3]
```diff
@@ -138,36 +145,41 @@ def _get_methylseq(dmr, bam_file_path: str, k: int, methyl_caller: str):
...
 def finetune_data_generate(
         f_dmr: str,
         output_dir: str,
         f_ref: str,
         sc_dataset: str = None,
-        input_file: str = None, 
+        input_file: str = None,
         n_mers: int = 3,
-        split_ratio: float = 1.0,
+        split_ratios: List[float] = [0.8, 0.1, 0.1],
```
[the remark this answers is not in the document]
Comment (peterdays) [reply src/methylbert/data/finetune_data_generate.py:172]: Hmm, I see! We should make it flexible for both cases, what do you think?

File: README.md [hunk 4]
```diff
@@ -16,18 +16,19 @@ bioRxiv 2023.10.29.564590; doi: https://doi.org/10.1101/2023.10.29.564590
 ## Installation
 _MethylBERT_ runs most stably with __Python=3.9__
```
[the remark this answers is not in the document]
Comment (peterdays) [reply README.md:17]: no, you are right! I've worked with python 3.11 only

File: src/methylbert/data/finetune_data_generate.py (hunk 1 above)
Comment (peterdays) [reply src/methylbert/data/finetune_data_generate.py:179]: Exactly! I find it more user friendly this way, one can have a custom read extraction different from the original one without the need to fork the package

File: src/methylbert/trainer.py (hunk 2 above)
Comment (peterdays) [reply src/methylbert/trainer.py:596]: You are right, I've only worked with gradient accumulation of 1, so I didn't notice that! It's up to you, don't find it necessary to add that division 🤔
